tiledb/api/src/query_arrow/fields.rs

```diff
@@ -0,0 +1,171 @@
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use arrow::array as aa;
+use arrow::datatypes as adt;
+
+use super::QueryBuilder;
+
+/// Default field capacity is 10MiB
+const DEFAULT_CAPACITY: usize = 1024 * 1024 * 10;
+
+#[derive(Debug, Default)]
+pub enum QueryField {
+    #[default]
+    Default,
+    WithCapacity(usize),
+    WithCapacityAndType(usize, adt::DataType),
+    WithType(adt::DataType),
+    Buffer(Arc<dyn aa::Array>),
+}
+
+impl QueryField {
+    pub fn capacity(&self) -> Option<usize> {
+        match self {
+            Self::Default => Some(DEFAULT_CAPACITY),
+            Self::WithCapacity(capacity) => Some(*capacity),
+            Self::WithCapacityAndType(capacity, _) => Some(*capacity),
+            Self::WithType(_) => Some(DEFAULT_CAPACITY),
+            Self::Buffer(_) => None,
+        }
+    }
+
+    pub fn target_type(&self) -> Option<adt::DataType> {
+        match self {
+            Self::Default => None,
+            Self::WithCapacity(_) => None,
+            Self::WithCapacityAndType(_, dtype) => Some(dtype.clone()),
+            Self::WithType(dtype) => Some(dtype.clone()),
+            Self::Buffer(array) => Some(array.data_type().clone()),
+        }
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct QueryFields {
+    pub fields: HashMap<String, QueryField>,
+}
+
+impl QueryFields {
+    pub fn insert<S: Into<String>>(&mut self, name: S, field: QueryField) {
+        let name: String = name.into();
+        self.fields.insert(name.clone(), field);
+    }
+}
+
+pub struct QueryFieldsBuilder {
+    fields: QueryFields,
+}
+
+impl QueryFieldsBuilder {
+    pub fn new() -> Self {
+        Self {
+            fields: Default::default(),
+        }
+    }
+
+    pub fn build(self) -> QueryFields {
+        self.fields
+    }
+
+    pub fn field(mut self, name: &str) -> Self {
+        self.fields.insert(name, Default::default());
+        self
+    }
+
+    pub fn field_with_buffer(
+        mut self,
+        name: &str,
+        buffer: Arc<dyn aa::Array>,
+    ) -> Self {
+        self.fields.insert(name, QueryField::Buffer(buffer));
+        self
+    }
+
+    pub fn field_with_capacity(mut self, name: &str, capacity: usize) -> Self {
+        self.fields.insert(name, QueryField::WithCapacity(capacity));
+        self
+    }
+
+    pub fn field_with_capacity_and_type(
+        mut self,
+        name: &str,
+        capacity: usize,
+        dtype: adt::DataType,
+    ) -> Self {
+        self.fields
+            .insert(name, QueryField::WithCapacityAndType(capacity, dtype));
+        self
+    }
+
+    pub fn field_with_type(mut self, name: &str, dtype: adt::DataType) -> Self {
+        self.fields.insert(name, QueryField::WithType(dtype));
+        self
+    }
+}
+
+pub struct QueryFieldsBuilderForQuery {
```

Comment:
Needs a doc comment.  If this one is "ForQuery" then what's the first one for?

Reply:
I infer that the answer is that this can add fields to an existing query

Reply:
First off, totes agree that we're gonna need better documentation stuff. Though this is the API stuff I'm not 100% sold on. A lot of this was just written in "make the interesting bits possible" mode so it's just enough to work. Though I think the fields thing in particular could use some improvement now that I've used it a bit.

The basic idea here was to support two ways of interacting with the QueryBuilder for adding fields:

```rust
    let mut query = QueryBuilder::new(array, QueryType::Read)
        .start_fields()
        .field("foo")
        .field("bar")
        .end_fields()
        .build()?;
```

```rust
    let fields = QueryFields::default()
        .field("foo")
        .field("bar")
        .build();

    let mut query = QueryBuilder::new(array, QueryType::Read)
        .with_fields(fields)
        .build()?;
```

And amusingly, I've used both of those in the examples. The first one for writing example code is useful enough, though I wouldn't be sad to lose it in the grand scheme of things. The second approach is used in the `reading_incomplete_arrow.rs` example because it fits with the way I landed for buffer reallocation.